### ISSUE TITLE
Error creating project with sample content

### ISSUE THREAD

**Issue by pq** [flutter/flutter#28896]
_From @akireyev on February 27, 2019 18:29_

## Steps to Reproduce

* Run a 'New Flutter Project' wizard in AS.
* Select 'generate sample content' with 'RaisedButton'

Expected: a new project is created.
Actual: an error in the log
```
Error creating project
Download failed.
URL: https://master-docs-flutter-io.firebaseapp.com/snippets/material.RaisedButton.dart
Error: 404 Not Found
```

## Version info

[√] Flutter (Channel beta, v1.1.8, on Microsoft Windows [Version 10.0.17134.619], locale en-US)
    • Flutter version 1.1.8 at ...
    • Framework revision 985ccb6d14 (7 weeks ago), 2019-01-08 13:45:55 -0800
    • Engine revision 7112b72cc2
    • Dart version 2.1.1 (build 2.1.1-dev.0.1 ec86471ccc)

[√] Android Studio (version 3.3)
    • Android Studio at D:\code\tools\JetBrains Toolbox\apps\AndroidStudio\ch-0\182.5264788
    • Flutter plugin version 33.3.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

_Copied from original issue: flutter/flutter-intellij#3236_

**Comment by pq** [comment]
Hi @akireyev.  Thanks for the report.  It looks like you have a fairly old Flutter SDK.  Could you update and try again?

**Comment by pq** [comment]
_From @akireyev on February 28, 2019 19:30_

Same result with
[√] Flutter (Channel beta, v1.2.1, on Microsoft Windows [Version 10.0.17134.619], locale en-US)
    • Flutter version 1.2.1 at ...
    • Framework revision 8661d8aecd (2 weeks ago), 2019-02-14 19:19:53 -0800
    • Engine revision 3757390fa4
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

**Comment by pq** [comment]
Thanks for confirming @akireyev.  We'll take a look. 👍

**Comment by pq** [comment]
Sure enough, it repros from the command-line too.

It looks like the sample ids in https://docs.flutter.io/snippets/index.json aren't valid anymore?

```
[~/src/tmp] $ flutter create --sample=material.Scaffold .
Download failed.
URL: https://master-docs.flutter.io/snippets/material.Scaffold.dart
Error: 404 Not Found
```

cc @gspencergoog @DanTup

**Comment by pq** [comment]
_From @DanTup on March 2, 2019 14:20_

The master version of that file has numbers suffixed into the IDs:

https://master-docs.flutter.io/snippets/index.json

So the issue here is that the IDEs don't know which host `flutter create` is going to use (I think we both just assumed the stable host is fine, and there wouldn't be breaking changes). The logic Flutter uses is checking which channel it's on, though it feels a bit clunky for the editor to start checking that.

Maybe we could make a `flutter create --list` so that IDEs can get the same JSON that matches what `flutter create` supports? It can just fetch the JSON and spit it out, but use the same internal logic for checking the hostname to avoid it leaking out?

**Comment by pq** [comment]
> Maybe we could make a flutter create --list so that IDEs can get the same JSON that matches what flutter create supports? It can just fetch the JSON and spit it out, but use the same internal logic for checking the hostname to avoid it leaking out?

I really like this idea.  I recall we discussed something like this at the outset (motivated by just this kind of worry).

@gspencergoog does that sound reasonable to you?

**Comment by pq** [comment]
_From @DanTup on March 4, 2019 9:32_

> I recall we discussed something like this at the outset (motivated by just this kind of worry).

I think there are actually two similar issues:

1. The users Flutter version being out-of-sync with the latest stable (which means the line numbers etc. may be incorrect)
2. The editor being out-of-sync with which hostname flutter tools will use

Having something like `flutter create --list` fixes the second one (breaking changes between stable/master won't cause issues like this), though sadly not the first one.

**Comment by pq** [comment]
> Having something like flutter create --list fixes the second one (breaking changes between stable/master won't cause issues like this), though sadly not the first one.

Could the output be structured so as to contain that data too?  (Maybe if `--machine` is passed in too.)

Regardless, this is looking to be more of a `flutter_tools` issue.  In which case, should we move this over to the Flutter repo?

**Comment by pq** [comment]
_From @DanTup on March 5, 2019 8:47_

> Could the output be structured so as to contain that data too?

Which data? If you mean the line numbers etc., they'd still potentially be wrong, because they come from the web, not from the local SDK. Bundling them inside the SDK would fix this, but from previous discussions I think may be complicated (we'd need to move snippet generation from docs to Flutter?).

> this is looking to be more of a `flutter_tools` issue

Yeah, I think so. I don't think editors should need to know about (or duplicate) the internal logic used to pick the hostname, so I think a fix should come from `flutter_tools` and then the editors updated as required.

**Comment by pq** [comment]
_From @gspencergoog on March 5, 2019 15:55_

Sure, if checking the channel isn't something you can do easily, then it makes sense to ask the flutter command to give you the JSON index.  It could just as easily give you the channel, but I agree that it doesn't make sense to duplicate the logic.

I did change the naming scheme for the snippets, but I didn't think that would affect anything because the new names (with the snippet indices) appear in the JSON index. I had to add an index, since there can be multiple examples tied to a single dartdoc location. I still think it shouldn't affect anything as long as you're fetching the right JSON data.

**Comment by pq** [comment]
_From @DanTup on March 5, 2019 17:26_

> I still think it shouldn't affect anything as long as you're fetching the right JSON data.

Yep, that's right. We're not fetching the right data because we didn't duplicate the logic for the hostname, so if you're on `master` there's a mismatch.

Moving to something like `flutter create --list` makes the fact that it goes to the server an implementation detail that the editors no longer care about (at least, when talking to a new SDK that supports it) which feels nicer. It also makes it easier to change this (eg. if we embedded them in the SDK) in future.

@pq I suspect you have a fancy way to move issues (there's a bot? I don't know how it works), so I'll leave that to you.

**Comment by pq** [comment]
> I suspect you have a fancy way to move issues (there's a bot? I don't know how it works), so I'll leave that to you.

⚡️ 

@DanTup: is this something you're interested in looking into?

**Comment by DanTup** [comment]
This feels like a trick question 😄

But sure, I can try and take a look a little later in the week. Do we just want `flutter create --list-samples` to dump a flat list of IDs, and `flutter create --list-samples --machine` to just dump the whole JSON file verbatim?

**Comment by DanTup** [comment]
(as a reminder of a possible spanner, it'll be possible for `flutter create --list-samples --machine` to output non-JSON and/or take a long time if it needs to download an SDK, or has an upgrade banner!)

**Comment by pq** [comment]
Yes.  Thanks for the reminder about the "first run" gotcha.  This gets us elsewhere too so it's best we use this as an opportunity to tidy up our handling a bit. 👍 

> But sure, I can try and take a look a little later in the week.

If you have time/interest, awesome.  And thanks!

**Comment by gspencergoog** [comment]
Couldn't you just supply an output location for the JSON to avoid the console output problem?

```bash
flutter create --list-samples=output.json --machine
```

Alternatively, we probably should be writing all that "downloading" "building flutter tool..." nonsense to the stderr anyhow.

**Comment by DanTup** [comment]
> Couldn't you just supply an output location for the JSON to avoid the console output problem?

Yeah, that might be an idea, though doesn't help the "may take a long time" bit - we just need to be sure we're not locking the UI or something for the duration (it's not so much a problem as a reminder that we should consider it when wiring to the UI).

> Alternatively, we probably should be writing all that "downloading" "building flutter tool..." nonsense to the stderr anyhow.

I think that'd complicate things just in different ways. We often collect stderr and show it to the user (to aid troubleshooting if something may have gone wrong). I think it'd be nice to have opt-out flags for some things (eg. `--no-update-notification`, `--no-download`) - though some of it would need to be handled in the shell scripts.

It'd also be nice if when we did have downloads we got the progress as JSON. In VS Code we detect the "Downloading..." line and show our own progress indicator until Flutter delivers what we expect. It works ok, but it'd be nice if we gave the same fidelity that the console does. That's probably for another day though :-)
